// ngx-resize-observer/esm2020/lib/ngx-resize-observer.module.mjs
import { NgModule } from '@angular/core';
import { NgxResizeObserverDirective } from './ngx-resize-observer.directive';
import { NgxResizeObserverService } from './ngx-resize-observer.service';
import * as i0 from "@angular/core";
export class NgxResizeObserverModule {
}
/** @nocollapse */ NgxResizeObserverModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.4", ngImport: i0, type: NgxResizeObserverModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
/** @nocollapse */ NgxResizeObserverModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.0.4", ngImport: i0, type: NgxResizeObserverModule, declarations: [NgxResizeObserverDirective], exports: [NgxResizeObserverDirective] });
/** @nocollapse */ NgxResizeObserverModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.0.4", ngImport: i0, type: NgxResizeObserverModule, providers: [NgxResizeObserverService] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.4", ngImport: i0, type: NgxResizeObserverModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgxResizeObserverDirective],
                    exports: [NgxResizeObserverDirective],
                    providers: [NgxResizeObserverService]
                }]
        }] });
