// ngx-resize-observer/esm2020/lib/ngx-resize-observer.service.mjs
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class NgxResizeObserverService {
    constructor(ngZone) {
        this.ngZone = ngZone;
        this.count = 0;
        this.elementMap = new Map();
        this.observer = null;
    }
    ngOnDestroy() {
        if (this.observer) {
            this.clearObserver();
        }
    }
    observe(element, callback, boxModel) {
        if (!this.observer) {
            this.observer = new ResizeObserver(resizes => {
                for (const resize of resizes) {
                    const cb = this.elementMap.get(resize.target);
                    if (cb) {
                        this.ngZone.run(() => {
                            cb(resize);
                        });
                    }
                }
            });
        }
        if (boxModel === 'border-box') {
            this.observer.observe(element, {
                box: 'border-box'
            });
        }
        else {
            this.observer.observe(element);
        }
        this.count += 1;
        this.elementMap.set(element, callback);
    }
    unobserve(element) {
        const cb = this.elementMap.get(element);
        if (cb && this.observer) {
            this.observer.unobserve(element);
            this.elementMap.delete(element);
            this.count -= 1;
            if (this.count === 0) {
                this.clearObserver();
            }
        }
    }
    clearObserver() {
        if (this.observer) {
            this.observer.disconnect();
        }
        this.observer = null;
        this.count = 0;
        this.elementMap = new Map();
    }
}
/** @nocollapse */ NgxResizeObserverService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.4", ngImport: i0, type: NgxResizeObserverService, deps: [{ token: i0.NgZone }], target: i0.ɵɵFactoryTarget.Injectable });
/** @nocollapse */ NgxResizeObserverService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.0.4", ngImport: i0, type: NgxResizeObserverService });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.4", ngImport: i0, type: NgxResizeObserverService, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i0.NgZone }]; } });
