// ngx-resize-observer/esm2020/lib/ngx-resize-observer.directive.mjs
import { Directive, EventEmitter, Input, Output } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "./ngx-resize-observer.service";
export class NgxResizeObserverDirective {
    constructor(elementRef, ngxResizeObserverService) {
        this.elementRef = elementRef;
        this.ngxResizeObserverService = ngxResizeObserverService;
        this.resizeBoxModel = '';
        this.onResize = new EventEmitter();
        this.observing = false;
    }
    ngAfterViewInit() {
        this.observe();
    }
    ngOnChanges(changes) {
        if (this.observing && (changes['resizeBoxModel'] || changes['onResize'])) {
            this.unobserve();
            this.observe();
        }
    }
    ngOnDestroy() {
        this.unobserve();
    }
    observe() {
        if (!this.observing) {
            this.ngxResizeObserverService.observe(this.elementRef.nativeElement, resize => this.onResize.emit(resize), this.resizeBoxModel);
            this.observing = true;
        }
    }
    unobserve() {
        if (this.observing) {
            this.ngxResizeObserverService.unobserve(this.elementRef.nativeElement);
            this.observing = false;
        }
    }
}
/** @nocollapse */ NgxResizeObserverDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.4", ngImport: i0, type: NgxResizeObserverDirective, deps: [{ token: i0.ElementRef }, { token: i1.NgxResizeObserverService }], target: i0.ɵɵFactoryTarget.Directive });
/** @nocollapse */ NgxResizeObserverDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.0.4", type: NgxResizeObserverDirective, selector: "[onResize]", inputs: { resizeBoxModel: "resizeBoxModel" }, outputs: { onResize: "onResize" }, usesOnChanges: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.4", ngImport: i0, type: NgxResizeObserverDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[onResize]'
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }, { type: i1.NgxResizeObserverService }]; }, propDecorators: { resizeBoxModel: [{
                type: Input
            }], onResize: [{
                type: Output
            }] } });
